core/src/main/kotlin/xtdb/indexer/LogProcessor.kt

```diff
@@ -36,7 +37,34 @@ class LogProcessor(
         private val LOG = LogProcessor::class.logger
     }
 
-    private val watchers = Watchers(liveIndex.latestCompletedTx?.txId ?: -1)
+    private val currentEpoch = log.currentEpoch
+
+    private fun offsetToTxId(offset: LogOffset): Long {
+        require(currentEpoch < epochLimit) { "Epoch value ($currentEpoch) exceeds the limit ($epochLimit)" }
+        require(offset < offsetLimit) { "Offset value ($offset) exceeds the limit ($offsetLimit)" }
+
+        return (currentEpoch.toLong() shl 48) + offset
+    }
+
+    private fun txIdToEpoch(txId: Long): Int {
```

Comment:
prefer expression bodies where possible (IntelliJ will give you this as an option if you Alt+Enter)

`private fun txIdToEpoch(txId: Long) = (txId shr 48).toInt()`